src/dragonball/src/dbs_pci/src/vfio.rs

```diff
@@ -0,0 +1,1918 @@
+/// which then gets added to the PCI bus.
+pub struct VfioPciDevice<C: PciSystemContext> {
+    pub(crate) id: u8,
+    pub(crate) vm_fd: Arc<VmFd>,
+    state: Mutex<VfioPciDeviceState<C>>,
+}
+
+///  VfioPciDevice destructor
+///
+/// Unregister regions before drop VfioPciDevice, otherwise the device
+/// fd's mmapped memory will not released.
+impl<C: PciSystemContext> Drop for VfioPciDevice<C> {
+    fn drop(&mut self) {
+        let mut state = self.state();
+        let _ = state.unregister_regions(&self.vm_fd);
+    }
+}
+
+impl<C: PciSystemContext> VfioPciDevice<C> {
+    /// Constructs a new Vfio Pci device for the given Vfio device
+    #[allow(clippy::too_many_arguments)]
+    pub fn create(
+        id: u8,
+        path: String,
+        bus: Weak<PciBus>,
+        device: VfioDevice,
+        context: Weak<C>,
+        vm_fd: Arc<VmFd>,
+        vendor_device_id: u32,
+        clique_id: Option<u8>,
+        vfio_container: Arc<VfioContainer>,
+    ) -> Result<Self> {
+        let device = Arc::new(device);
+        let mut state = VfioPciDeviceState::new(
+            path,
+            device,
+            Weak::clone(&bus),
+            context,
+            vendor_device_id,
+            clique_id,
+            vfio_container,
+        )?;
+
+        state.probe_regions(bus)?;
+        state.parse_capabilities();
+        state.parse_legacy_irq();
+        state.fixup_msix_region();
+
+        Ok(VfioPciDevice {
+            id,
+            vm_fd,
+            state: Mutex::new(state),
+        })
+    }
+
+    /// Get resource requirements of the VFIO PCI device.
+    pub fn get_resource_requirements(&self, requests: &mut Vec<ResourceConstraint>) {
+        let state = self.state();
+        if !state.sys_constraints.is_empty() {
+            let mut constraints = state.sys_constraints.clone();
+            requests.append(&mut constraints);
+        }
+    }
+
+    /// Because when routing MSI, KVM needs to know the device id of the device corresponding to the MSI.
+    /// Therefore, when VfioPciDevice is activated, we will record its device id in its irq_manager
+    /// structure for routing MSI.
+    #[cfg(target_arch = "aarch64")]
+    fn set_device_id(&self, state: &mut VfioPciDeviceState<C>) {
+        let device_id = Some(self.id as u32);
+        state
+            .interrupt
+            .irq_manager
+            .as_mut()
+            .unwrap()
+            .set_device_id(device_id);
+    }
+
+    pub fn activate(&self, device: Weak<dyn DeviceIo>, resources: DeviceResources) -> Result<()> {
+        let mut state = self.state();
+
+        if resources.len() == 0 {
+            return Err(VfioPciError::InvalidResources);
+        }
+
+        let mut all_kvm_slot = resources.get_kvm_mem_slots();
+
+        let mut reg_idx = 0;
+        while reg_idx < state.regions.len() {
+            if state.regions[reg_idx].mappable() {
+                state.regions[reg_idx].mmaps[0].slot = all_kvm_slot.pop().unwrap();
```

Comment:
is it safe to unwrap? same below

Reply:
add a new Err `KvmSlotNotFound` for this and delete `unwrap`.